### ISSUE TITLE
A11y keyboard focus control in the List Component

### ISSUE THREAD

**Issue by DaveBest99** [angular/components#3682]
#### Bug, feature request, or proposal:
Bug - List component has no content structure markup.

#### What is the expected behavior?
A screen reader user should be able to distinguish between the List Items.

#### What is the current behavior?
The demo List Component displays a list of 5 items with each item having multiple lines (image, heading, paragraph). The md-list has no focusable item, and the md-list-item objects have no defined structure; Thus a screen reader user cannot distinguish between item content.

#### What are the steps to reproduce?
See below.

Providing a Plunker (or similar) is the *best* way to get the team to see your issue.
Plunker template: https://goo.gl/DlHd6U

#### What is the use-case or motivation for changing an existing behavior?
Create a good keyboard user experience.

#### Which versions of Angular, Material, OS, browsers are affected?
Using screen reader JAWS17 with browsers Firefox, Internet Explorer 11, and Google Chrome, on Windows10.

#### Is there anything else we should know?
Screen reader user experience follows:

1. Press Tab key to the List element in the Tab Panel.
The screen reader skips over the Tab Panel and moves focus to the page footer. A md-list object could use the tabindex=0 to include it in the focus order.

2. Use the arrow keys to move screen reader focus to the Tab Panel List element.
JAWS says "list of 5 items", and braille display shows "list of 5 items".
As the user presses Down Arrow JAWS speaks the line text and the type of element, but does not indicate which lines belong to which List Item. All list items are merged into a single block of text.

3. The author must manage focus order for screen readers to understand content context.

3.1 The md-list uses role=list and md-list-item uses role=listitem, but no keyboard controls are provided.
Try including ARIA Supported States and Properties: aria-level, aria-posinset, aria-setsize.

3.2 Another possible solution may be to use the role=group property.
Use a group to form logical collection of items in a widget. when a group is used in the context of list, authors MUST limit its children to listitem elements.

3.3 Another solution may be in using the role=grid and role=gridcell properties.
The grid pattern can be used to group a set of interactive elements, such as links, buttons, or checkboxes. This may be useful for large md-list objects with interactive elements within each md-list-item.

See:
The Roles Model | Accessible Rich Internet Applications (WAI-ARIA) 1.0
http://w3c.github.io/aria-practices/#grid

Demo Angular Material - Component List 
https://material.angular.io/components/component/list

**Comment by jelbourn** [comment]
I can also see the item distinction problem with VoiceOver + Chrome. Some things I think may be problematic:
* Having content elements such as `<h4>` and `<p>` inside of `role="listitem"`, which may only be intended to contain phrasing content. 
* The icons in the list items should be `aria-hidden`

I don't believe the list as a whole should have any tab stops in general, as it's not interactive content. Because the spec specifically mentions lists are not interactive, I think we are using `role="list"` incorrectly in general (especially for nav-list). 

We need more research/experimentation into the appropriate way to have this kind of list that contains complicated content within each list item.

**Comment by DaveBest99** [comment]
It is not clear why this section of page content is enclosed in a List element, other than possibly for visual layout. The WCAG criteria does not define the List element structure, but it is not recommended to enclude Headings and Paragraphs within a List element without proper HTML markup structure. The icons can be ARIA hidden if they offer no additional information that is already provided by text.

 

The List element may not be interactive, but it must be in the focus order for keyboard users. Whether there are Tab stops within the List element (for each List Item) may depend upon the size of the List, as it would be very time consuming and tedious to have to Down Arrow key through the entire list just to get to the List Item desired. If each List Item contains multiple HTML elements, then keyboard navigation would become a challenge. It is one thing to comply with WCAG criteria, but another to be user friendly. Possibly List Items could be enclosed in <article> with Headings and Paragraphs. The important thing is to establish a consistent pattern for the Angular brand style. There are many different techniques to implement WCAG success criteria’s.

 

David

**Comment by mckenzielong** [comment]
Believe this is related... MatNavList also has issues with focus.  See https://stackblitz.com/edit/angular-6fsgmx?file=app%2Flist-sections-example.html and tab through the elements.
